src/OrleansRuntime/Messaging/IncomingMessageAcceptor.cs

```diff
@@ -235,153 +330,176 @@ private static void AcceptCallback(IAsyncResult result)
-                {
-                    // There was a network error. We need to get a new accepting socket and re-issue an accept before we continue.
-                    // Close the old socket and open a new one
-                    ima.Log.Warn(ErrorCode.MessagingEndAcceptSocketException, "Exception on accepting socket during EndAccept", ex);
-                    ima.RestartAcceptingSocket();
-                    return;
-                }
+                        // Finally, process the incoming request:
+                        // Prep the socket so it will reset on close
+                        sock.LingerState = receiveLingerOption;
 
-                if (ima.Log.IsVerbose3) ima.Log.Verbose3("Received a connection from {0}", sock.RemoteEndPoint);
+                        // Add the socket to the open socket collection
+                        if (ima.RecordOpenedSocket(sock))
+                        {
+                            // Get the socket for the accepted client connection and put it into the 
+                            // ReadEventArg object user token.
+                            var readEventArgs = GetSocketReceiveAsyncEventArgs(sock);
 
-                // Finally, process the incoming request:
-                // Prep the socket so it will reset on close
-                sock.LingerState = new LingerOption(true, 0);
+                            StartReceiveAsync(sock, readEventArgs, ima);
+                        }
+                        else
+                        {
+                            ima.SafeCloseSocket(sock);
+                        }
 
-                // Add the socket to the open socket collection
-                if (ima.RecordOpenedSocket(sock))
-                {
-                    // And set up the asynch receive
-                    var rcc = new ReceiveCallbackContext(sock, ima);
-                    try
+                    }
+                    catch (SocketException ex)
                     {
-                        rcc.BeginReceive(ReceiveCallback);
+                        Log.Warn(ErrorCode.Messaging_ExceptionReceiveAsync, "Error when processing data received from {0}:\r\n{1}", "Q", ex, sock.RemoteEndPoint);
                     }
-                    catch (Exception exception)
+                    catch (Exception ex)
                     {
-                        var socketException = exception as SocketException;
-                        ima.Log.Warn(ErrorCode.Messaging_IMA_NewBeginReceiveException,
-                            String.Format("Exception on new socket during BeginReceive with RemoteEndPoint {0}: {1}",
-                                socketException != null ? socketException.SocketErrorCode.ToString() : "", rcc.RemoteEndPoint), exception);
-                        ima.SafeCloseSocket(sock);
+                        this.Log.Warn(ErrorCode.Messaging_ExceptionReceiveAsync, "Exception trying to process accept from endpoint ", ex);
+                        throw;
```

Comment:
The catch few line below will handle it and restart the socket. Added restarting in line.